File: api/js/src/validator/plugins/plugin-manager.ts
import { RuleRegistry } from "../context/rule-registry.js";
import type { ValidationPlugin } from "../types.js";

export class PluginManager {
  private readonly registry: RuleRegistry;
  private readonly loaded = new Set<string>();

  constructor(registry: RuleRegistry) {
    this.registry = registry;
  }

  use(plugin: ValidationPlugin): void {
    if (this.loaded.has(plugin.name)) {
      return;
    }
    if (plugin.rules) {
      for (const rule of plugin.rules) {
        this.registry.registerRule(rule);
      }
    }
    plugin.setup?.(this.registry);
    this.loaded.add(plugin.name);
  }
}

File: api/python/telegram_inline_keyboard_builder/validator/plugins/plugin_manager.py
from __future__ import annotations

from ..context.rule_registry import RuleRegistry
from ..types import ValidationPlugin


class PluginManager:
    def __init__(self, registry: RuleRegistry) -> None:
        self._registry = registry
        self._loaded: set[str] = set()

    def use(self, plugin: ValidationPlugin) -> None:
        if plugin["name"] in self._loaded:
            return
        if "rules" in plugin:
            for rule in plugin["rules"]:
                self._registry.register_rule(rule)
        if "setup" in plugin:
            plugin["setup"](self._registry)
        self._loaded.add(plugin["name"])
